File: oceanlk-next/models/AuditLog.ts
import mongoose, { Schema, Document, Model } from 'mongoose';

export interface IAuditLog extends Document {
    username: string;
    action: string; // CREATE, UPDATE, DELETE, LOGIN
    entityType: string;
    entityId?: string;
    details?: string;
    timestamp: Date;
}

const AuditLogSchema = new Schema<IAuditLog>({
    username: { type: String, required: true, index: true },
    action: { type: String, required: true },
    entityType: { type: String, required: true, index: true },
    entityId: { type: String, index: true },
    details: String,
    timestamp: { type: Date, default: Date.now, index: true }
}, {
    collection: 'audit_logs',
    // Capped collection: 10MB size limit or 5000 documents
    capped: { size: 10485760, max: 5000 }
});

const AuditLog: Model<IAuditLog> = mongoose.models.AuditLog || mongoose.model<IAuditLog>('AuditLog', AuditLogSchema);

export default AuditLog;

File: apps/backend/src/main/java/com/oceanlk/backend/model/AuditLog.java
package com.oceanlk.backend.model;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;

@Document(collection = "audit_logs")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class AuditLog {
    @Id
    private String id;

    private String username; // The admin who performed the action
    private String action; // CREATE, UPDATE, DELETE, LOGIN
    private String entityType; // JobOpportunity, MediaItem, etc.
    private String entityId; // ID of the affected entity
    private String details; // Optional description or change diff

    private LocalDateTime timestamp;

    public AuditLog(String username, String action, String entityType, String entityId, String details) {
        this.username = username;
        this.action = action;
        this.entityType = entityType;
        this.entityId = entityId;
        this.details = details;
        this.timestamp = LocalDateTime.now();
    }
}
